File: Assets/scripts/GameMain/CardSlot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;

public class CardSlot
{
    private string color;
    private List<Card> cards;
    public List<Card> Cards
    {
        get { return cards; }
    }

    public CardSlot(string color)
    {
        this.color = color;
        this.cards = new List<Card>();
    }

    public void Add(Card card)
    {
        cards.Add(card);
    }

    public int GetTotalPower()
    {
        int sum = 0;
        for (int i = 0; i < cards.Count; i ++)
        {
            Card item = cards.ElementAt(i);
            if (item.CanBattle())
            {
                sum += item.Power;
            }
        }
        return sum;
    }

    private Card GetLowestPower()
    {
        if (cards.Count > 0)
        {
            Card ret = null;
            int min = 9999;
            for(int i = 0; i < cards.Count; i ++)
            {
                Card current = cards.ElementAt(i);
                if (current.CanBattle())
                {
                    int power = current.Power;
                    if (power < min)
                    {
                        min = power;
                        ret = current;
                    }
                }
            }
            return ret;
        }
        else
        {
            return null;
        }
    }

    internal List<Card> TakeDamage(int num)
    {
        List<Card> killed = new List<Card>();
        if (cards.Count > 0)
        {
            bool exhausted = false;
            while (num > 0 && !exhausted)
            {
                Card current = GetLowestPower();
                if (current == null)
                {
                    exhausted = true;
                }
                else
                {
                    if (current.Power > num)
                    {
                        // aborbed the damage
                        current.Power -= num;
                        num = 0;
                    }
                    else
                    {
                        // killed by this damage
                        num -= current.Power;
                        current.Power = 0;
                        cards.Remove(current);
                        killed.Add(current);
                    }
                }
            }
        }
        return killed;
    }

    internal void CheckBuffsEndTurn(Player caster)
    {
        foreach(Card card in cards)
        {
            card.CheckBuffsEndTurn(caster);
        }
    }

    internal List<Tuple<int, int>> GetModifiers(bool attacker)
    {
        List<Tuple<int, int>> res = new List<Tuple<int, int>>();
        for (int i = 0; i < cards.Count; i ++)
        {
            Card item = cards.ElementAt(i);
            if (item.CanBattle())
            {
                Assembly assembly = Assembly.GetExecutingAssembly();
                Type type = null;
                try
                {
                    type = assembly.GetTypes().First(t => t.Name == "Card" + item.Id);
                }
                catch (InvalidOperationException ioe)
                {
                    Debug.LogWarning(ioe);
                }
                if (type != null)
                {
                    CardAbility obj = (CardAbility)Activator.CreateInstance(type);
                    obj.Card = item;
                    int mod = obj.GetAttackModifier(attacker);
                    if (mod != 0)
                    {
                        res.Add(new Tuple<int, int>(item.Id, mod));
                    }
                }
            }
        }
        return res;
    }
}

File: Server/src/CardSlot.js
const Events = require('./constants/Events');
const CardAbility = require('./cardAbilities/CardAbility');

class CardSlot {

    constructor(color, id, player) {
        this.id = id;
        this.owner = player;
        this.color = color;
        this.attacks = 1;
        this.cards = [];
    }

    add(card) {
        this.cards.push(card);
    }

    getCards() {
        return this.cards;
    }

    getTotalPower() {
        let sum = 0;
        for (let i = 0; i < this.cards.length; i++) {
            const item = this.cards[i];
            if (item.canBattle()) {
                sum += item.getPower();
            }
        }
        return sum;
    }

    restoreSlotAttackCharges() {
        this.attacks = 1;
    }

    consumeCharge() {
        this.attacks -= 1;
    }

    getAttackCharges() {
        return this.attacks;
    }

    getLowestPower() {
        if (this.cards.length > 0) {
            let ret = null;
            let min = 9999;
            for(let i = 0; i < this.cards.length; i ++) {
                const current = this.cards[i];
                if (current.canBattle()) {
                    let power = current.getPower();
                    if (power < min) {
                        min = power;
                        ret = current;
                    }
                }
            }
            return ret;
        } else {
            return null;
        }
    }

    takeDamage(num) {
        const killed = [];
        const touched = [];
        if (this.cards.length > 0) {
            let exhausted = false;
            while (num > 0 && !exhausted) {
                const current = this.getLowestPower();
                if (current == null) {
                    exhausted = true;
                } else {
                    if (current.getPower() > num) {
                        // aborbed the damage
                        current.takeDamage(num);
                        touched.push(current);
                        num = 0;
                    } else {
                        // killed by this damage
                        num -= current.getPower();
                        current.takeDamage(num);
                        this.cards.splice(this.cards.indexOf(current), 1);
                        killed.push(current);
                    }
                }
            }
        }
        return {
            touched: touched,
            killed: killed
        };
    }

    checkBuffsEndTurn(caster) {
        let ret = [];
        for (let i = 0; i < this.cards.length; i++) {
            const resp = this.cards[i].checkBuffsEndTurn(caster);
            if (resp.length > 0) {
                ret = ret.concat(resp);
                ret = ret.concat([{
                    ename: Events.UPDATE_SLOT_POWER,
                    payload: {
                        name: this.owner.getName(),
                        slot: this.id,
                        power: this.getTotalPower()
                    }
                }]);
            }
        }
        return ret;
    }

    getAllCardData() {
        const res = [];
        for (let i = 0; i < this.cards.length; i++) {
            const item = this.cards[i];
            if (item.canBattle()) {
                res.push(item.getData());
            }
        }
        return res;
    }

    getId() {
        return this.id;
    }

    getModifiers(isAttacker) {
        const res = [];
        let sum = 0;
        for (let i = 0; i < this.cards.length; i ++) {
            const item = this.cards[i];
            if (item.canBattle()) {
                const className = "Card" + item.getId();
                if (CardAbility.hasOwnProperty(className)) {
                    const obj = new CardAbility[className](item);
                    if (CardAbility[className].prototype.hasOwnProperty("getBattleModifier")) {
                        let mod = obj.getBattleModifier(isAttacker);
                        if (mod != 0) {
                            res.push({
                                guid: item.getGuid(),
                                mod: mod
                            });
                            sum += mod;
                        }
                    }
                }
            }
        }
        return {
            mods: res,
            sum: sum
        };
    }
}

module.exports = CardSlot;